### PR TITLE
Slack alerting for DAG failures

### LINKED ISSUE #214: Add DAG alerting in Slack
## Problem

Unless someone is watching Airflow directly, we won't know when DAG failures occur.

## Description

We can set up a Slack app/API client to inform us of when DAGs fail, so we have an opportunity to address them as soon as they occur.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #214 by @AetherUnbound

## Description

This PR adds an `on_failure_callback` method to the `slack` module which will (in certain conditions) propagate that failure message to Slack. This failure callback has been added to most DAGs.

image

I also added an `environment` Airflow Variable to the environment-defined variables. This value can be helpful for making decisions about alerting/what actions to take in a given situation depending on whether the instance is being run locally (`dev`) or in production (`prod`).

Lastly I made some minor tweaks to the other `slack` tests to make mock access a bit clearer.

## Testing Instructions

`just test`

If you'd like to test the notification, you'll need to:
1. Set up the Slack hook (defined in the `.env` file)
1. Add an Airflow Variable `force_slack_alert` with value `true`
1. Instigate a DAG failure on one of the altered DAGs

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: openverse_catalog/dags/common/slack.py [hunk 1]
```diff
@@ -213,3 +218,51 @@ def send_message(
...
+        return
+
+    # Get relevant info
+    ti = context["task_instance"]
+    execution_date = context["execution_date"]
+    exception: Optional[Exception] = context.get("exception")
+    exception_message = ""
+
+    if exception:
+        # Forgo the alert on upstream failures
```
Comment (zackkrida) [inline openverse_catalog/dags/common/slack.py:250]: nice

**Review by zackkrida — APPROVED** [review]
Nice test coverage and addition to the codebase!

File: tests/dags/common/test_slack.py [hunk 2]
```diff
@@ -268,8 +270,8 @@ def test_send_with_context(http_hook_mock):
 def test_send_fails(http_hook_mock):
     s = SlackMessage()
     # Cause an exception within the raise_for_status call
-    http_hook_mock.return_value.run.return_value.raise_for_status.side_effect = (
-        Exception("HTTP Error 666")
+    http_hook_mock.run.return_value.raise_for_status.side_effect = Exception(
+        "HTTP Error 666"
```
Comment (zackkrida) [inline tests/dags/common/test_slack.py:274]: Very, very trival but I wonder if this should be clearer that it's a fake value for testing. It's probably fine 😅

File: openverse_catalog/dags/common/slack.py (hunk 1 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/slack.py:250]: If we don't do this, every task downstream of the primary failure _also_ issues an alert :joy:

File: tests/dags/common/test_slack.py (hunk 2 above)
Comment (AetherUnbound) [reply tests/dags/common/test_slack.py:274]: Ha, I can reword this message :slightly_smiling_face:

File: openverse_catalog/dags/common/slack.py (hunk 1 above)
Comment (zackkrida) [reply openverse_catalog/dags/common/slack.py:250]: ha, I had a feeling...

**Review by krysal — APPROVED** [review]
LGTM!

**Pull request merged**